### PR TITLE
[Bugfix] Reduce Triton TILE_SIZE on Blackwell for large head_size with float32

### PR DESCRIPTION
**Pull request by kitaekatt**
## Summary

Blackwell GPUs (SM 120) have a shared memory limit of 101,376 bytes per block. When using `head_size=256` with `float32` (`element_size=4`), `TILE_SIZE=32` requires ~117KB, exceeding this limit:

```
triton.runtime.errors.OutOfResources: out of resource: shared memory,
Required: 116992, Hardware limit: 101376
```

This affects Gemma2 and Gemma3 GGUF models which use `head_size=256` and require `float32` dtype for correct inference on Blackwell (float16 produces degenerate logits, bfloat16 has GGUF dequant precision issues).

## Fix

Add a Blackwell detection guard in `_get_tile_size()` that reduces `TILE_SIZE` to 16 (~66KB) when:
- Running on Blackwell (compute capability 12.x)
- `head_size=256`
- `element_size=4` (float32)

## Testing

Tested with `bartowski/gemma-2-2b-it-GGUF` on RTX 5090 (SM 120):
- **Before**: `OutOfResources` crash on first inference request
- **After**: Correct output at 373 tok/s with `TRITON_ATTN` backend, `dtype=float32`

## Shared Memory Budget

| Config | TILE_SIZE | Approx Shared Mem | Fits? |
|--------|-----------|-------------------|-------|
| head_size=256, float32 | 32 | ~117KB | No (limit: 101KB) |
| head_size=256, float32 | 16 | ~66KB | Yes |
| head_size=256, float16 | 32 | ~66KB | Yes (not affected) |
| head_size=128, float32 | 32 | ~66KB | Yes (not affected) |

### CODE REVIEW

**Comment by kitaekatt** [comment]
Hi @tdoublep — this has been sitting without review for a while. Just rebased on latest `upstream/main` (merge commit), so the branch should now be mergeable. Quick summary: head_dim=256 + float32 at the default TILE_SIZE=32 overflows SM 120's 101KB shared-mem limit (~117KB required). Guards the Blackwell path to TILE_SIZE=16 (~66KB). Required for Gemma2 GGUF on Blackwell. Rebased on latest main.

Would appreciate a look when you have cycles. Thanks!
